Treat autocompletions as tokens with editing boundaries

brandonpayton opened the pull request:
## Description
This PR updates the `<Autocomplete>` component to insert autocompletions as tokens.

Completions consisting entirely of text are inserted as styled, editable tokens. This satisfies the common completer use cases of user mentions (@-user), tags (#-tag), and cross posting site mentions (+site).

Here's an example of editable tokens:
![first-completer-type-example](https://user-images.githubusercontent.com/530877/42016409-44b5550a-7a60-11e8-91ec-f0e61745324a.gif)

Completions containing HTML are inserted as non-editable tokens. This supports the creation of completers that insert arbitrary HTML content.

Here's an example of a non-editable token created from [this completer](https://gist.github.com/brandonpayton/01b75c55ce4d329d534b1d3a68dc728f):
![second-completer-type-example](https://user-images.githubusercontent.com/530877/41986081-f3685830-79e9-11e8-833e-009499a51aed.gif)

Closes #6573.

## How has this been tested?
Tested manually in Chrome, Firefox, and Safari on macOS and in IE11 and Edge on Windows.

I also added an e2e test for basic token insertion.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

brandonpayton committed:
Wrap autocompletions as tokens

iseulde commented:
So the user is not allowed to edit e.g. a mention? Why not? This PR does not clarify why this change is needed.

noisysocks commented on components/autocomplete/index.js:
```diff
@@ -226,15 +226,39 @@ export class Autocomplete extends Component {
 		range.deleteContents();
+
+		/*
+		 * Add non-breaking space after a completion because:
```
Doesn't this mean, though, that there will never be a break in between a mention and the next word? This could result in awkward-looking text flows.

For example, this mention at the end of the line causes `stuff @editor done` to go onto a new line:

<img width="773" alt="screen shot 2018-05-07 at 11 29 25" src="https://user-images.githubusercontent.com/612155/39681052-83d5cb5a-51ea-11e8-8eb1-448217afa53d.png">

Here's how it would look without non-breaking spaces:

<img width="757" alt="screen shot 2018-05-07 at 11 32 08" src="https://user-images.githubusercontent.com/612155/39681048-7f9c532e-51ea-11e8-9553-5149867937ea.png">

Maybe it's a compromise we're OK with. Just checking 🙂

brandonpayton replied:
@noisysocks It is definitely a compromise. In my testing, I was unable to add a regular space and programmatically place the cursor afterward. This appears to be the issue or at least related:
https://bugs.chromium.org/p/chromium/issues/detail?id=428313

I thought about using zero-width non-breaking spaces as TinyMCE seems to do to move the cursor around [inline boundaries](https://codepen.io/tinymce/pen/EwLYBM), making it easier to specifically target and edit things like link text. But the character is not always removed by the editor, and it's invisible nature can cause problems. For example, in Firefox, when I backspace from one paragraph block into another with a trailing text node containing a zero-width non-breaking space, the cursor is not placed at the end of the text but rather one character before the end.

Inserting a visible non-breaking space is the best I've come up with to place the cursor and have editing work as expected after inserting an autocomplete token. I'm not totally OK with it but figured we could live with it.

I'd love to hear other ideas though.

brandonpayton commented:
> So the user is not allowed to edit e.g. a mention? Why not? This PR does not clarify why this change is needed.

Thanks for asking, @iseulde. I should have included my reasoning. I updated the PR with more of my reasoning and would love to hear what you think.

brandonpayton commented:
I spoke with Matías, and he'd like these to be editable. I'm working on an update now.

brandonpayton commented:
I updated this PR to make the tokens editable with inline boundaries, a generic `autocomplete-token` class for general styling, and an `autocomplete-token-<completer-name>` class for styling tokens from specific completers.

brandonpayton commented:
@karmatosed, is there a particular style you'd like to see applied to autocomplete tokens either in general or specifically (like styles for the tokens inserted for @ user mentions)?

brandonpayton commented:
When discussing integrating Gutenberg with P2, @mtias mentioned that it might be good to highlight xpost tokens like:
![image](https://user-images.githubusercontent.com/530877/39879040-b0c937fa-542e-11e8-8dc0-af743a696212.png)

My impression is that he wanted them highlighted in some way. It's not a blocker for this PR though. I'm adding classes to the tokens, and we can add styles for them later. Based on your input, I'll leave them unstyled for now. Thank you!

iseulde commented on components/autocomplete/index.js:
```diff
@@ -225,22 +225,44 @@ export class Autocomplete extends Component {
+
+		tokenWrapper.innerHTML = renderToString( replacement );
+
+		range.insertNode( tokenWrapper );
```
Wondering if we should move this to dom utils and keep all range logic there...

brandonpayton replied:
I feel hesitant to move this to the dom utils because I dislike the `&nbsp;` compromise so much :-P

But if we were to move this to dom utils, what do you imagine we'd create there? Something like this?

```typescript
// Please feel free to suggest improved name.
export function replaceRangeWithToken( range : Range, value : <string|WPElement>, tokenClasses : Array<string> ) {
  const tokenWrapper = document.createElement( 'span' );
  tokenClasses.forEach( ( c ) => tokenWrapper.classList.add( c ) );
  tokenWrapper.innerHTML = renderToString( replacement );
  
  range.insertNode( tokenWrapper );
  range.setStartAfter( tokenWrapper );
  range.deleteContents();
  
  // <comment explaining why nbsp>
  tokenWrapper.parentNode.insertBefore(
  	document.createTextNode( '\u00A0' ),
  	tokenWrapper.nextSibling
  );
  
  const selection = window.getSelection();
  selection.removeAllRanges();
  
  const newCursorPosition = document.createRange();
  newCursorPosition.setStartAfter( tokenWrapper.nextSibling );
  selection.addRange( newCursorPosition );
}
```

I feel like the above may be too specific to be valuable for reuse, but it _would_ be nice to keep Range-related logic in the dom utils.

iseulde commented:
Would this be a good candidate for an e2e test?

brandonpayton commented:
> Would this be a good candidate for an e2e test?

Seems like an excellent candidate. Thanks, I'll work on one.

brandonpayton commented:
I added an e2e test for token insertion, but it is _really_ clunky. I'm going to give my mind a break and then return to see if I can clean it up.

brandonpayton commented:
I updated the e2e test. It's still not nearly as readable as I'd like, but I think it's better.

mtias commented:
There seems to be a bug where you edit an existing mention, delete a letter or two, get the auto-suggestion popover, select the name again and you end up with two:

![image](https://user-images.githubusercontent.com/548849/40304417-fb0853e8-5cf6-11e8-879d-4964f3eb4594.png)

mtias commented on components/autocomplete/index.js:
```diff
@@ -225,22 +225,44 @@ export class Autocomplete extends Component {
-		}
+	insertCompletion( range, replacement, completerName ) {
+		// Wrap completions so we can treat them as tokens with editing boundaries.
+		const tokenWrapper = document.createElement( 'span' );
```
Don't we have a native tinymce API for this?

iseulde replied:
There are TinyMCE APIs for this, but I'm guessing this component shouldn't depend on TinyMCE? Only the `RichText` component is aware of it. This is similar to writing flow which has tried to avoid using TinyMCE APIs. If we want we could pass the `editor` instance though.

mtias commented on components/autocomplete/style.scss:
```diff
@@ -34,3 +34,7 @@
 	}
 }
+
+.autocomplete-token[data-mce-selected] {
```
cc @jasmussen for color choice and whether we should show the background at all times, not just on selection (similar to the code inline format).

jasmussen replied:
I think @karmatosed is looking at this now. My thoughts are these:

- We should definitely leverage, if we can, the inline boundaries for tokens as well. 
- Inline boundaries, like bold, italic, links, should feel "ephemeral" and disappear when you're not interacting with the boundary. Subtle colors and no borders make them feel ephemeral.
- The tokens should feel more substantial, and should probably _keep_ their boundaries even when you're not interacting with them. A subtle background plus border can help make them feel like "chips", which I think would make them feel more substantial, which seems a good thing.

brandonpayton commented:
Thanks, @mtias.

> There seems to be a bug where you edit an existing mention, delete a letter or two, get the auto-suggestion popover, select the name again and you end up with two

My mind also sent this to me over the weekend, realizing I hadn't added any logic to prevent it.

I believe the preferred behavior is that the completion menu displays when editing a token but only replaces the existing token. I'll work on a fix like this, but please let me know if you'd like a different behavior.

brandonpayton commented:
There are some challenges to displaying the completion menu when editing tokens since completions may have arbitrary content and structure. Here are some reasons:
1. Completions are not required to include their `triggerPrefix` (e.g., how user mentions include the '@' character). For example, here's an abbreviations completer that doesn't include its trigger prefix in the completion:
https://gist.github.com/brandonpayton/01b75c55ce4d329d534b1d3a68dc728f
2. The content of a completion is not guaranteed to be a query for the option used to generate the completion. In other words, if the completion text is different than what the user typed, we can't display the same list of options when editing. For example, a synonym completer would take one word and allow selection of a synonym X, but when editing, the original word is lost, and instead, we might propose synonyms for X.

There are straightforward cases like user mentions (@-user), tags (#-tag), and xposts (+blogname). They contain their trigger prefix, and the completion is also a valid, specific query for the option that generated the completion. We can handle these cases and know how to show a completion menu when editing.

Other cases may require more work to handle due to the challenges mentioned above.

Some brainstorming on solution or workaround:
* Store the query used to generate a completion and use that somehow when editing.
* Allow marking completers with an `editableCompletions` flag to indicate whether completions can be edited. Document the requirements for a truly editable completion. The straightforward cases could be editable, and the more complicated cases could be non-editable.
* Update the completer design to only support the simple cases.

My intent is not to expand our scope but to consider what is currently possible with today's completer design so we can properly handle editing.

@mtias, do you have any thoughts on this? My impression is that you may have a simpler vision for completers than what I described above.

youknowriad commented:
I'm punting as it doesn't seem ready yet for 3.0.

brandonpayton committed:
Handle editing tokens

brandonpayton commented:
> You: Design wise we need to make sure they stand out a little more. Let's explore this:

> Me: Is this design specifically for user completions or for autocomplete tokens in general?

I should have read your example more closely. It looks like this is specifically for user mentions since your example shows tags and cross posts without styles. Sorry about that.

brandonpayton committed:
Insert zero-width non-breaking space instead

After more testing, it is a better UX to insert
a zero-width space because:
1. When a space is inserted to place the cursor
afterward, the user occassionally hits backspace
to remove it, and doing so places the cursor
back within the token (at least in Chrome 66).
If we use zero-width non-breaking spaces,
the user can simply add a space when desired
and a reduced chance of accidentally re-placing
the cursor within the token immediately
after inserting it.
2. TinyMCE seems to handle ZWNBSPs well at
token boundaries when moving the cursor
with the keyboard.

brandonpayton committed:
Update user completion style

brandonpayton commented:
This has been updated to fix the nested insertion bug mentioned by @mtias.

I also noticed a negative user experience around inserting spaces along with tokens. If the user doesn't want the space and hits backspace to remove it, their cursor is placed back within the token rather than outside as we intended. Because of this, I ended up inserting a zero-width non-breaking space (ZWNBSP) instead so we can place the cursor outside the token after insertion. This reduces the possibility the user will accidentally put their cursor back within the token after inserting it, and TinyMCE appears to handle ZWNBSP's well when moving the cursor in and out of token boundaries using the keyboard.

I added initial style updates for the user completer based on the design above. I used our color variables and tried to pick grays near the ones used in [the example](https://github.com/WordPress/gutenberg/pull/6577#issuecomment-392867603).

<img width="690" alt="screen shot 2018-06-10 at 3 41 35 pm" src="https://user-images.githubusercontent.com/530877/41207266-9db116cc-6cc7-11e8-8eb9-72d6a31a821e.png">

brandonpayton committed:
Update the e2e test for zero-width space

karmatosed commented:
Thanks for all the hardwork on this @brandonpayton, let's see about getting you some answers to your questions.

I first think there's been a little misconception, there isn't meant to be a UI for this. I think that can clear some of this up if we avoid that as a focus.

What I am suggesting is a default color in the grey. Then the third party can configure other color options. This isn't something the user would interact with.

I think this answers most of your questions but if you have anything else lets tackle it with this information in mind.

brandonpayton committed:
Style tweaks after discussion

brandonpayton commented:
Thanks, @karmatosed. That clears up many of my questions and matches the current state of this PR. The last question is whether the autocomplete tokens should somehow be integrated with or use the same registration as the tokens in our [inline images PR](https://github.com/WordPress/gutenberg/pull/6959). That is probably one for @mtias.

I updated the padding on the autocomplete tokens after talking with you. What do you think of this?
<img width="698" alt="screen shot 2018-06-12 at 3 58 15 pm" src="https://user-images.githubusercontent.com/530877/41321678-eed1abfa-6e59-11e8-8b94-f846be012bed.png">

mtias commented:
> The last question is whether the autocomplete tokens should somehow be integrated with or use the same registration as the tokens in our inline images PR.

@brandonpayton yes. The goal is that, when caret is inline, you'd see registered tokens in the inserter (mentions, etc). But this is a separate task and should happen after inline images is merged.

brandonpayton committed:
Add a couple more why comments

brandonpayton commented:
Thanks, @iseulde and @mtias. The token is currently being wrapped in a `span` with a generic `autocomplete-token` class and a completer-specific class for styling, so I think we're good there. My concern was that I was creating something that wouldn't integrate well with our greater vision for tokens, but it sounds like it will be OK.

I believe this is ready for review.

iseulde commented on components/autocomplete/index.js:
```diff
@@ -225,22 +225,85 @@ export class Autocomplete extends Component {
+		const element = isTextNode( possibleTokenChild ) ?
+			possibleTokenChild.parentNode :
+			possibleTokenChild;
+		const withinToken = element.matches( '.autocomplete-token, .autocomplete-token *' );
```
Is the dependency defined in this file? See also #7159.

brandonpayton replied:
Thanks for the reminder. I missed that.

I added import of 'element-closest' to add the `Element#matches` polyfill for now, but I'm hoping to work on a better way of registering and including polyfills later this week.

iseulde commented:
I'll have a closer look at the code as soon as possible!

brandonpayton committed:
Add 'element-closest' dep for `Element#matches`

mtias commented:
If I write an exact match but don't choose it from the auto-completer, it doesn't wrap it in the format boundary. We should probably do it for exact matches.

![image](https://user-images.githubusercontent.com/548849/41604531-c5f2ec4a-73df-11e8-8bb8-9e35b8518f58.png)

brandonpayton commented:
@mtias, I agree that is awkward, but we don't currently have a way to match autocomplete options with strings. We just have the autocomplete option UI that can be selected. I'll give some thought to how we can solve this. In the meantime, is that a blocker for this PR?

brandonpayton commented:
Another issue is that the query for options is async, so even if we had a way to determine an exact match, we don't really know until the options are received which may be later than when the input for "@test2" is received. Async replacement may be jarring.

For determining whether entered text could be replaced with a completion token, we may be able to match against option keywords, but there is not one exact string that can be matched per completer option.

brandonpayton commented:
Due to the async nature of option retrieval, we could schedule an option query when the user follows "@test2" with a space and then replace if:
1. The text still exists when the options are received.
2. there is an option with a keyword matching "test2".

I don't love the experience I imagine with this solution, but having the text remain unwrapped as a token is also unfortunate.

iseulde commented:
@mtias I'm going to look at the RichText state this week and with that it will be possible.

iseulde commented on components/autocomplete/index.js:
```diff
@@ -225,22 +226,85 @@ export class Autocomplete extends Component {
+	insertCompletion( range, replacement, completerName ) {
+		// Wrap completions so we can treat them as tokens with editing boundaries.
+		const tokenWrapper = document.createElement( 'span' );
+		tokenWrapper.innerHTML = renderToString( replacement );
```
What if the replacement is an element? Should we maybe only wrap gin a span when it's a string?

brandonpayton replied:
Thanks, @iseulde, I think that's a great suggestion. If a completer wants to insert custom HTML, it could, and we wouldn't interfere by wrapping it in a token span. I am out of time today but plan to address this in the morning.

brandonpayton replied:
Actually, I'm not sure what to do here because, even if we don't wrap with a span, we probably want editing boundaries for any autocompletion. What do you think?

Should we add a class to the element if a completer returns one? What if a completer returns a fragment? This may have been the line of thinking I followed to wrap everything in a span, though I understand not all elements are valid children of a span.

brandonpayton replied:
After sleeping on this, my thinking is:
* String completions should be wrapped as editable tokens. I believe this is the simple autocompletion experience we often expect.
* Element completions should be marked as read-only since it is arbitrary content and may be awkward to edit properly.

What do you think, @mtias?

mtias replied:
It sounds like there could be two types of completers, but given lack of examples of the second kind, we could optimize for the cases we have now.

brandonpayton replied:
Here is an example of the second kind where a completer is used to insert an acronym:
https://gist.github.com/brandonpayton/01b75c55ce4d329d534b1d3a68dc728f

![second-completer-type-example](https://user-images.githubusercontent.com/530877/41986081-f3685830-79e9-11e8-833e-009499a51aed.gif)

In this case, the completion does not include the trigger prefix, and its `<abbr>` element has a `title` attribute that cannot be edited via the RichText component. My take is that we should not wrap these in the same way but should mark them as readonly so the editor can remove but not edit them. Does that example help?

mtias replied:
Sounds good to me.

iseulde commented on components/autocomplete/index.js:
```diff
@@ -225,22 +226,85 @@ export class Autocomplete extends Component {
+			 * If we're within an existing token, we want to replace it rather
+			 * than inserting a completion within a completion.
+			 */
+			existingTokenWrapper.parentNode.replaceChild( tokenWrapper, existingTokenWrapper );
```
I think all these DOM manipulations need to we wrapped in [`editor.undoManager.transact`](https://www.tinymce.com/docs/api/tinymce/tinymce.undomanager/#transact) so that we are sure this is one history record. Becomes harder to do though as this component is not aware of the TinyMCE editor instance.

iseulde commented:
Looking at this code, I'm again also realising that we could really use a better editable state for this and an easy to use API for applying formats/tokens/elements/..., setting selection, and matching the text. Looking into it again this week. In the meantime this seems fine to merge.

brandonpayton commented:
Thanks for your review, @iseulde!

I had time to work on @mtias's request to insert a token if the user spaces after typing an exact option match, but it has a couple of bugs I'm looking into. In the morning, I'm hoping to clean that up and follow up on other review comments.

brandonpayton committed:
Insert token when spacing after exact match

brandonpayton commented:
> If I write an exact match but don't choose it from the auto-completer, it doesn't wrap it in the format boundary. We should probably do it for exact matches.

Hi @mtias, I've pushed an update to fix this. What is your take on [this review conversation](https://github.com/WordPress/gutenberg/pull/6577#discussion_r196740969)? After that question is addressed, I believe this is ready for consideration.

brandonpayton committed:
Make non-text completions readonly

brandonpayton commented:
I pushed an update to treat non-text autocompletions as readonly tokens. The result is:

1. Simple, textual completions are wrapped as editable tokens and styled as @karmatosed suggested.
2. Completions containing elements or more than one DOM nodes are treated as readonly tokens with no default styling. This is because we cannot fully edit arbitrary HTML content, and it is cleaner to restrict them to being added or removed.

@iseulde would you be up for re-reviewing this? I believe I agree with you that we need a better, more generic facility for these kinds of insertions. This PR is not my favorite code, but I believe it is reasonable in function.

brandonpayton commented:
I am looking at making documentation updates since there are now two ways for us to treat completions.

iseulde commented on components/autocomplete/index.js:
```diff
@@ -147,6 +148,17 @@ function isTextNode( node ) {
+ * @return {boolean} True if the node is an element.
+ */
+function isElement( node ) {
+	return node !== null && node.nodeType === window.Node.ELEMENT_NODE;
```
Other files have this usually as a browser dependency. Same for above for text.

brandonpayton replied:
Do you mean simply checking `nodeType`? That sounds right. I went with the pattern I saw in creating `isTextNode`, but it's unnecessary.

iseulde commented on editor/components/rich-text/tinymce.js:
```diff
@@ -163,14 +163,15 @@ export default class TinyMCE extends Component {
 
-		settings.plugins.push( 'paste' );
+		settings.plugins.push( 'paste', 'noneditable' );
+		settings.noneditable_noneditable_class = 'readonly-token';
```
So we're making this non editable again?

brandonpayton replied:
The latest change treats completions one of two ways:
1. Simple, textual completions are treated as editable tokens.
2. Completions with an element or more than one node are treated as readonly because they are arbitrary content we cannot be sure to support in the editor.

Does that make sense?

iseulde replied:
Unsure. Perhaps this should be up to the author. E.g. inserting an image shouldn't be make non editable? Or is this for fragments?

brandonpayton replied:
I am also not 100% sure. Currently, the idea was to make all non-text completions non-editable. My thinking comes from working with examples like the abbreviation completer shown in this PR description and thinking it is strange to allow editing because the user can edit text children but not attributes of the inserted `<abbr>` element. And with my limited experience with contenteditable, a user can get into strange places editing arbitrary HTML. So my first inclination was to pick a default that avoids that.

But you have more knowledge and wisdom about these things, and I'm interested in your take on my concerns.

brandonpayton committed:
Clean up recent additions

brandonpayton committed:
Document different inserted token treatments

notnownikki commented on components/autocomplete/README.md:
````diff
@@ -107,7 +114,29 @@ Whether to apply debouncing for the autocompleter. Set to true to enable debounc
+Here is a contrived completer for feelings. It yields editable tokens like "!resolute".
+
+```jsx
+const fruitCompleter = {
````
Should this be `feelingCompleter`?

brandonpayton replied:
Possibly :)  Thanks for catching that.

iseulde commented on components/autocomplete/index.js:
```diff
@@ -560,6 +659,24 @@ export class Autocomplete extends Component {
+		if (
+			event.keyCode === SPACE &&
+			! ( ctrlKey || shiftKey || altKey || metaKey ) &&
+			open && suppress !== open.idx
```
From reading this piece of code, I have no idea what this means.

iseulde commented on components/autocomplete/index.js:
```diff
@@ -560,6 +659,24 @@ export class Autocomplete extends Component {
+		) {
+			// Insert a completion when the user spaces after typing an exact option match.
+			const exactMatchSearch = new RegExp( '^' + escapeRegExp( query ) + '$', 'i' );
+			const wasOptions = this.state[ 'options_' + open.idx ];
```
What's `wasOptions`?

iseulde commented:
I'm encountering a bug when I insert a mention inside formatting. The mention gets inserted at the start.

<img width="230" alt="screen shot 2018-06-28 at 17 44 16" src="https://user-images.githubusercontent.com/4710635/42045120-06350e10-7afb-11e8-89f7-acff19e6287d.png">

iseulde commented:
Also, when I change the mention, and there is no user by it, the styling stays.

iseulde commented:
There seems to be a point at which the pop up is still visible after deleting the mention with backspace.

<img width="358" alt="screen shot 2018-06-28 at 17 48 33" src="https://user-images.githubusercontent.com/4710635/42045402-bee3a0d4-7afb-11e8-924d-6901e5cf2add.png">

brandonpayton committed:
Updates in response to code review

brandonpayton commented:
> I'm encountering a bug when I insert a mention inside formatting. The mention gets inserted at the start.

Thanks, @iseulde. In what browser are you seeing this? I am unable to reproduce in Chrome.

brandonpayton commented:
> Also, when I change the mention, and there is no user by it, the styling stays.

Since tokens are simply inserted HTML, I was thinking that the user could do whatever they want with them when they are editable. It's not styled based on a pattern match, though that was the first thing I wanted when starting to work on this. Is there a different approach you would suggest?

brandonpayton commented:
I've reproduced the bug with inserting the mention inside formatting but haven't discovered how to reproduce it reliably. I'm continuing to poke at it.

brandonpayton committed:
Fix missing input event for backspace

brandonpayton commented:
I pushed 0c0f4708e37dac39d65a7f247b309147db6b5dc0 to stop the completion menu from lingering after deleting a token with backspace.

The remaining bug is incorrectly inserting tokens inside formatting as mentioned [here](https://github.com/WordPress/gutenberg/pull/6577#issuecomment-401080624).

brandonpayton commented:
The incorrect insertion placement appears to be caused by a DOM Range referring to a text node that has been replaced. The range first refers to the text node, then later refers to the text node's parent with its start and end offsets zeroed. I am not yet sure how to fix this but am working on it.

brandonpayton commented:
> The incorrect insertion placement appears to be caused by a DOM Range referring to a text node that has been replaced. The range first refers to the text node, then later refers to the text node's parent with its start and end offsets zeroed. I am not yet sure how to fix this but am working on it.

It looks like the solution needs to be identifying the range on-demand since the editor occasionally replaces text nodes and changes our saved range. I likely need to wait until Monday to make this change but plan to work on it first thing.

brandonpayton committed:
Fix incorrect insertion due to stale DOM range

brandonpayton commented:
Hi @iseulde, I believe this is ready for another review.

Undo remains a bit wonky related to your comment [here](https://github.com/WordPress/gutenberg/pull/6577#discussion_r196741730):
> I think all these DOM manipulations need to we wrapped in [`editor.undoManager.transact`](https://www.tinymce.com/docs/api/tinymce/tinymce.undomanager/#transact) so that we are sure this is one history record. Becomes harder to do though as this component is not aware of the TinyMCE editor instance.

Do you see a reasonable way to fix this with the current `Autocomplete` component design? It definitely feels like there should be more direct integration with TinyMCE rather than direct DOM manipulation here. `Autocomplete` is written in such a way that it assumes `contenteditable` anyway.

It would also be good to get your feedback on my responses to your comments [here](https://github.com/WordPress/gutenberg/pull/6577#discussion_r198952039) and [here](https://github.com/WordPress/gutenberg/pull/6577#issuecomment-401163511).

Thank you for your time on this.

brandonpayton commented:
I played with adding an optional `transactInsertion` prop to `<Autocomplete>` to facilitate integration with an undo stack. It seems to work well, but I need to test a bit more and document before pushing.

brandonpayton committed:
Make autocompletions undoable

brandonpayton commented:
I added a change to effectively wrap insertions with `editor.undoManager.transact`. The undo experience is better but not perfect. Perhaps it is good enough for a first version of autocomplete tokens.

brandonpayton commented:
Hi @azaozz, would you be up for reviewing my changes here? Reading the PR description and conversation starting [here](https://github.com/WordPress/gutenberg/pull/6577#issuecomment-401918980) should be a sufficient background. I understand there may not be bandwidth at the moment but wanted to ask. Thanks for your consideration. :)

brandonpayton commented:
I'm closing this as it would be stronger and cleaner to implement autocomplete tokens based on @iseulde's rich text structure updates.
